### PR TITLE
Module4 demo

### PR DESCRIPTION
**Pull request by SiarheiKrystseu**

### CODE REVIEW

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/App.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.epam.esm.security.AuditorAwareImpl;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
+import org.springframework.data.domain.AuditorAware;
+import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
+
+@SpringBootApplication
+@EnableJpaAuditing(auditorAwareRef = "auditorAware")
```
Comment (sap-answerhub-team) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/App.java:11]: Move logic for audit to separate config class and keep main class clean

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/config/WebSecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                .antMatchers(HttpMethod.GET, "/gift-certificates/certificates").permitAll()
+                .antMatchers(HttpMethod.GET, "/gift-certificates/certificates/{id}").permitAll()
+                .antMatchers(HttpMethod.GET, "/gift-certificates/certificates/search").permitAll()
+                .antMatchers(HttpMethod.GET, "/gift-certificates/tags").permitAll()
+                .antMatchers(HttpMethod.GET, "/gift-certificates/tags/{id}").permitAll()
+                .antMatchers(HttpMethod.GET, "/gift-certificates/tags/popular").permitAll()
+                .antMatchers(HttpMethod.GET, "/gift-certificates/login").permitAll()
+                .antMatchers(HttpMethod.POST, "/gift-certificates/login").permitAll()
+                .antMatchers(HttpMethod.GET, "/gift-certificates/signup").permitAll()
+                .antMatchers(HttpMethod.POST, "/gift-certificates/signup").permitAll()
```
Comment (sap-answerhub-team) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/config/WebSecurityConfig.java:62]: Optimize matchers: either use wildcard 
`antMatchers("/gift-certificates/**").permitAll()`
or use permitAll as default access method

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+        BindingResult result = new BeanPropertyBindingResult(query, "searchQuery");
+        searchValidator.validate(query, result);
+        if (result.hasErrors()) {
+            String brokenField = result.getFieldErrors().get(0).getField();
+            String errorCode = result.getFieldErrors().get(0).getCode();
+            throw new InvalidInputDataException(MessageFormat.format("Unexpected certificate''s field: {0}, error code: {1}",
+                    brokenField, errorCode));
+        }
+        Page<GiftCertificateDTO> certificateList
```
Comment (sap-answerhub-team) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java:196]: This is not a good practice to use "list" word in a variable name when it is not a List interface implementation.

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+     * @return ResponseEntity.
+     */
+    @PostMapping(path = "/certificates", consumes = "application/json", produces = "application/json")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    public ResponseEntity<GiftCertificateDTO> addGiftCertificate(@RequestBody @Valid GiftCertificateDTO giftCertificateDTO,
+                                                                 UriComponentsBuilder ucBuilder) {
+        Long certificateId = giftCertificateService.saveCertificate(giftCertificateDTO);
+        HttpHeaders headers = new HttpHeaders();
+        headers.setLocation(ucBuilder.path("/certificates/{id}").buildAndExpand(certificateId).toUri());
+        return new ResponseEntity<>(headers, HttpStatus.CREATED);
```
Comment (sap-answerhub-team) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java:84]: Return certificate object in response body. According to method signature result should be of ResponseEntity<GiftCertificateDTO> type

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/App.java (hunk 1 above)
Comment (SiarheiVarabei) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/App.java]: Move logic for audit to separate config class and keep main class clean

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/config/WebSecurityConfig.java (hunk 2 above)
Comment (SiarheiVarabei) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/config/WebSecurityConfig.java]: Optimize matchers: either use wildcard
`antMatchers("/gift-certificates/**").permitAll()`
or use permitAll as default access method

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java (hunk 4 above)
Comment (SiarheiVarabei) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java]: Return certificate object in response body. According to method signature result should be of _ResponseEntity<GiftCertificateDTO>_ type

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java (hunk 3 above)
Comment (SiarheiVarabei) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java]: This is not a good practice to use "list" word in a variable name when it is not a List interface implementation.

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/OrderController.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+     *
+     * @param id Order id.
+     * @return Order instance.
+     */
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping(value = "/orders/{id}", produces = {"application/hal+json"})
+    @PreAuthorize("hasAnyRole('ROLE_USER', 'ROLE_ADMIN') && @authorizationComponentImpl.isUsersOrder(principal, #id)")
+    public OrderDTO findOrderById(@PathVariable Long id) {
+        OrderDTO orderDTO = orderService.getOrderById(id);
+        return addHateoasLinksToOrder(orderDTO);
```
Comment (SiarheiVarabei) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/OrderController.java:100]: This is a good idea to move HATEOAS related logic to separate method. Code looks clean and readable.

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/converter/OrderConverter.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor(onConstructor_ = @Autowired)
+public class OrderConverter {
+
+    private final GiftCertificateConverter giftCertificateConverter;
+    private final UserConverter userConverter;
+
+    public OrderDTO convertOrderDTOFromOrder(Order order) {
```
Comment (SiarheiVarabei) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/converter/OrderConverter.java]: Use simplified method names, for example 
```
OrderConverter.convertFromDto()
OrderConverter.convertFromEntity()
```

File: module4/gift-certificate-system/api/src/main/java/com/epam/esm/filter/JwtRequestFilter.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+
+    private final UserDetailsService userDetailsService;
+    private final JwtTokenProvider jwtTokenProvider;
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+            throws ServletException, IOException {
+
+        String username = (String) request.getAttribute(USERNAME_REQUEST_ATTRIBUTE);
```
Comment (SiarheiVarabei) [inline module4/gift-certificate-system/api/src/main/java/com/epam/esm/filter/JwtRequestFilter.java]: Username and all other attributes should be loaded from jwt token.

File: module4/gift-certificate-system/core/src/main/java/com/epam/esm/model/Role.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package com.epam.esm.model;
+
+import lombok.*;
+
+import javax.persistence.*;
+import java.util.Set;
+@Data
+@AllArgsConstructor()
+@NoArgsConstructor
```
Comment (SiarheiVarabei) [inline module4/gift-certificate-system/core/src/main/java/com/epam/esm/model/Role.java:9]: Why do you need so many constructors? @Data annotation includes @RequiredArgsConstructor that should add default constructor for this class. Also I can not see where do you use @AllArgsConstructor in code.

**Pull request merged**